pkg/api/norman/customization/gke/handler.go

```diff
@@ -0,0 +1,194 @@
+package gke
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/norman/api/access"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/util"
+	client "github.com/rancher/rancher/pkg/client/generated/management/v3"
+	v1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rancher/pkg/ref"
+	schema "github.com/rancher/rancher/pkg/schemas/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+)
+
+type Capabilities struct {
+	Credentials string `json:"credentials,omitempty"`
+	ProjectID   string `json:"projectId,omitempty"`
+	Zone        string `json:"zone,omitempty"`
+	Region      string `json:"region,omitempty"`
+}
+
+// GKE handler lists available resources in Google API
+type handler struct {
+	Action        string
+	schemas       *types.Schemas
+	secretsLister v1.SecretLister
+}
+
+func NewGKEHandler(scaledContext *config.ScaledContext) http.Handler {
+	return &handler{
+		schemas:       scaledContext.Schemas,
+		secretsLister: scaledContext.Core.Secrets(namespace.GlobalNamespace).Controller().Lister(),
+	}
+}
+
+func (h *handler) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+
+	writer.Header().Set("Content-Type", "application/json")
+
+	capa := &Capabilities{}
+
+	if credID := req.URL.Query().Get("cloudCredentialId"); credID != "" {
+		if err := h.getCloudCredential(req, capa, credID); err != nil {
+			util.ReturnHTTPError(writer, req, httperror.InvalidBodyContent.Status, err.Error())
+			return
+		}
+	} else if req.Method == http.MethodPost {
+		if err := h.getCredentialsFromBody(writer, req, capa); err != nil {
+			util.ReturnHTTPError(writer, req, httperror.InvalidBodyContent.Status, err.Error())
+			return
+		}
+	} else {
+		util.ReturnHTTPError(writer, req, httperror.Unauthorized.Status, "cannot access Google API without credentials to authenticate")
+	}
+
+	var serialized []byte
+	var errCode int
+	var err error
+
+	resourceType := mux.Vars(req)["resource"]
+
+	switch resourceType {
+	case "gkeMachineTypes":
+		if serialized, errCode, err = listMachineTypes(req.Context(), capa); err != nil {
+			logrus.Debugf("[gke-handler] error getting machine types: %v", err)
```

Comment:
Looks like we changed how the call gets returned. Specifically that postCheck isn't called anymore. Are we still writing necessary headers ?

Reply:
It works for gke v1 and gke v2 with Cloudcredentials

Reply:
can you check this @mjura

Reply:
@westlywright can you please judge on this ? do you know if HTTP error CODE is checked somewhere in UI? As I have checked and tested it is not, but please confirm. According to comment from @cmurphy we would like to use more proper HTTP error codes https://github.com/rancher/rancher/pull/31759#discussion_r603460276

Reply:
I have added back old error code response

Reply:
My comment before was about returning a 5xx server error for an invalid request, which wouldn't make sense from an API perspective and is also backwards incompatible with how it was before https://github.com/rancher/rancher/pull/31759/files#diff-d86bb7d4d3eda030b6a6ce369760bcf4f949ea0eda69b9d841f1f741c18a5d69L29

So http.StatusBadRequest is fine, that's how it was before and makes sense from an API usability perspective. http.InvalidOption would not have made any sense.